stack.hpp

```diff
@@ -16,17 +21,17 @@ class Stack {
     }
 
     void push(const T& element) {
-        if (position_ > capacity_) throw std::out_of_range("Not enough capacity");
-        storage_[position_++] = element;
+        if (position_ >= capacity_) throw std::out_of_range("Not enough capacity");
+        storage_[++position_] = element;
     }
 
     void push(T&& element) {
-        if (position_ > capacity_) throw std::out_of_range("Not enough capacity");
-        storage_[position_++] = std::move(element);
+        if (position_ >= capacity_) throw std::out_of_range("Not enough capacity");
+        storage_[++position_] = std::move(element);
```

Comment:
Pre-increment is wrong at insertion. This will cause the storage buffer to overrun.